File: BOJ/11725.cpp
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

int main(void) {
	int N;
	vector<vector<int>> adjacents;
	vector<int> parent;

	queue<int> visiting_queue;

	ios::sync_with_stdio(false);
	cin.tie(NULL);

	// get the number of nodes
	cin >> N;

	// init tree (index 0 is dummy)
	adjacents.resize(N + 1);

	// get all edges
	for (int i = 1; i < N; i++) {
		int p, q;
		cin >> p >> q;
		adjacents[p].push_back(q);
		adjacents[q].push_back(p);
	}

	// init parent (index 0 is dummy)
	parent.resize(N + 1, 0);

	// start from the root: node 1
	parent[1] = 1;
	visiting_queue.push(1);
	while (!visiting_queue.empty()) {
		// pop one
		int cur = visiting_queue.front();
		visiting_queue.pop();

		// visit all adjacent nodes
		for (int adj : adjacents[cur]) {
			if (parent[adj] == 0) {
				// if not visited, set parent and push to queue
				parent[adj] = cur;
				visiting_queue.push(adj);
			}
		}
	}

	// print parents
	for (int i = 2; i <= N; i++) {
		cout << parent[i] << '\n';
	}

	return 0;
}

File: BOJ/11725.ts
import * as fs from 'fs';

const inputManager: Generator<string, any, unknown> = (function *() {
  const stdin = '/dev/stdin';
  const lines = fs.readFileSync(stdin).toString().split('\n');
  for (const line of lines) yield line;
  return null;
})();
const nextLine = (): string => inputManager.next().value;
const nextLineNumbers = () => nextLine().split(' ').map(Number);

type Node = {
  id: number;
  parent: number | null;
  adjacents: Node[];
};

// init all nodes (index 0 is dummy)
const [N] = nextLineNumbers();
const nodes: Array<Node> = new Array(N + 1).fill(null).map((_, i) => ({
  id: i,
  parent: null,
  adjacents: [],
}));

// get all edges
for (let i = 0; i < N - 1; i++) {
  const [p, q] = nextLineNumbers();
  nodes[p].adjacents.push(nodes[q]);
  nodes[q].adjacents.push(nodes[p]);
}

// start from the root: node 1
nodes[1].parent = 1;
const visitingQueue = [nodes[1]];
while (visitingQueue.length) {
  // pop one
  const node = visitingQueue.shift()!;

  // visit all adjacent nodes
  // set parent and push to queue
  visitingQueue.push(
    ...node.adjacents.filter(node => node.parent === null).map(adj => {
      adj.parent = node.id;
      return adj;
    })
  );
}

// print parents
console.log(
  nodes.slice(2).map(node => node.parent).join('\n')
);
